### PR TITLE
Support LayerNormalization for opset 17

### PR DESCRIPTION
**Pull request by ShuaiShao93**
Issue: https://github.com/onnx/tensorflow-onnx/issues/2183

### CODE REVIEW

File: tf2onnx/onnx_opset/nn.py [hunk 1]
```diff
@@ -959,6 +959,92 @@ def version_11(cls, ctx, node, **kwargs):
...
+
+        inp_shape = ctx.get_shape(node.input[0])
+        inp_rank = len(inp_shape) if inp_shape is not None else None
+        scale_shape = ctx.get_shape(node.input[1])
+        mean_shape = ctx.get_shape(node.input[3])
+        var_shape = ctx.get_shape(node.input[4])
+        val_type = utils.map_onnx_to_numpy_type(ctx.get_dtype(node.input[1]))
+        is_training = node.get_attr_value('is_training', True)
+
+        if is_training and node.get_attr_value('exponential_avg_factor', 1.0) == 1.0:
```
Comment (fatcat-z) [inline tf2onnx/onnx_opset/nn.py:1013]: Thanks for your contributions!

Could you please create a common method where this loop could be refactored like [here](https://github.com/onnx/tensorflow-onnx/blob/b57fe915b3728fbf996b490170a374821dd8383a/tf2onnx/onnx_opset/nn.py#L835C9-L835C9)? So those different versions could call this common method and avoid duplicated code as much as possible.

Comment (ShuaiShao93) [reply tf2onnx/onnx_opset/nn.py]: Ah I didn't know that. Done.

Comment (fatcat-z) [reply tf2onnx/onnx_opset/nn.py]: Besides the pylint issue, looks like this change has broken some legacy tests, please also fix them. Thanks.

Comment (ShuaiShao93) [reply tf2onnx/onnx_opset/nn.py]: Thanks! I mainly use ONNX for TRT, and don't know much about ONNX runtime. But it seems the ONNX layernorm doesn't support automatic broadcasting on scale tensor. So I need to do it manually. But I realized there is no broadcast op in ONNX either. Can you suggest how to broadcast an ONNX tensor?

Comment (fatcat-z) [reply tf2onnx/onnx_opset/nn.py]: > Thanks! I mainly use ONNX for TRT, and don't know much about ONNX runtime. But it seems the ONNX layernorm doesn't support automatic broadcasting on scale tensor. So I need to do it manually. But I realized there is no broadcast op in ONNX either. Can you suggest how to broadcast an ONNX tensor?

Could you please follow the approach [here](https://github.com/onnx/tensorflow-onnx/blob/ca17b3c99eb37dc3f42c9be5a64620d247a24115/tf2onnx/onnx_opset/tensor.py#L4017C6-L4017C6) to try Expand op?

File: tf2onnx/onnx_opset/nn.py [hunk 2]
```diff
@@ -1011,25 +1015,46 @@ def version_6(cls, ctx, node, **kwargs):
...
-            dims_const = ctx.make_const(utils.make_name("axes_const"), np.array(dims, dtype=np.int64)).output[0]
-            reduce_dims = ctx.make_node("Gather", [inp_shape, dims_const]).output[0]
-            dims_product = GraphBuilder(ctx).make_reduce_prod({"data": reduce_dims, "axes": [0], "keepdims": False})
-            cnt_float = ctx.make_node("Cast", [dims_product], attr={'to': ctx.get_dtype(node.input[0])})
-
-            pop_var_squeezed = ctx.make_node("Div", [var_squeezed, cnt_float.output[0]]).output[0]
-            ctx.replace_inputs(node, node.input[:3] + [avg_squeezed, pop_var_squeezed])
+            if ctx.opset >= 17:
+                node.type = "LayerNormalization"
+                first_axis_to_norm = 2
```
Comment (aboubezari) [inline tf2onnx/onnx_opset/nn.py:1020]: Hey @ShuaiShao93, why did you add this change? Don't we have to extract the axis value from the axis attribute already in the TF op?

Comment (ShuaiShao93) [reply tf2onnx/onnx_opset/nn.py:1020]: I don't quite understand this very well either. I just saw line 1035 seems to assume axis=2 is first_axis_to_norm. And I remember I could verify the parity end to end.

Comment (ShuaiShao93) [reply tf2onnx/onnx_opset/nn.py:1020]: BTW a better alternative is probably making https://github.com/onnx/tensorflow-onnx/blob/main/tf2onnx/rewriter/layer_normalization_rewriter.py work, so that we can also get rid of a bunch of other ops added by TF
